File: glut_opengl_estudos/04_entrada_teclado_mouse.py
"""
Aula 04 — Interação com teclado e mouse

Conceitos:
- Callback de teclado (`glutKeyboardFunc`).
- Callback de mouse (`glutMouseFunc`).
- Estado global simples para aplicações interativas.

Atalhos:
- W/S: mover para cima/baixo
- A/D: mover para esquerda/direita
- R/G/B: trocar cor
- Botão esquerdo do mouse: reposicionar objeto
- ESC: sair

Como executar:
    python3 04_entrada_teclado_mouse.py
"""

from OpenGL.GL import (
    GL_COLOR_BUFFER_BIT,
    GL_QUADS,
    glBegin,
    glClear,
    glClearColor,
    glColor3f,
    glEnd,
    glLoadIdentity,
    glTranslatef,
    glVertex2f,
)
from OpenGL.GLUT import (
    GLUT_DOWN,
    GLUT_LEFT_BUTTON,
    GLUT_RGB,
    GLUT_SINGLE,
    GLUT_WINDOW_HEIGHT,
    GLUT_WINDOW_WIDTH,
    glutCreateWindow,
    glutDisplayFunc,
    glutGet,
    glutInit,
    glutInitDisplayMode,
    glutInitWindowSize,
    glutKeyboardFunc,
    glutMainLoop,
    glutMouseFunc,
    glutPostRedisplay,
)

position_x = 0.0
position_y = 0.0
current_color = (0.9, 0.4, 0.2)


def draw_square() -> None:
    """Desenha quadrado com centro na origem local."""
    glBegin(GL_QUADS)
    glVertex2f(-0.15, -0.15)
    glVertex2f(0.15, -0.15)
    glVertex2f(0.15, 0.15)
    glVertex2f(-0.15, 0.15)
    glEnd()


def display() -> None:
    """Renderiza quadrado conforme estado atual."""
    glClear(GL_COLOR_BUFFER_BIT)
    glLoadIdentity()

    glTranslatef(position_x, position_y, 0.0)
    glColor3f(*current_color)
    draw_square()


def keyboard(key: bytes, _: int, __: int) -> None:
    """Processa teclas e atualiza estado global."""
    global position_x, position_y, current_color

    step = 0.05
    if key == b"w":
        position_y += step
    elif key == b"s":
        position_y -= step
    elif key == b"a":
        position_x -= step
    elif key == b"d":
        position_x += step
    elif key == b"r":
        current_color = (1.0, 0.2, 0.2)
    elif key == b"g":
        current_color = (0.2, 1.0, 0.2)
    elif key == b"b":
        current_color = (0.2, 0.4, 1.0)
    elif key == b"\x1b":
        raise SystemExit("Encerrado pelo usuário (ESC).")

    glutPostRedisplay()


def mouse(button: int, state: int, x: int, y: int) -> None:
    """Converte coordenadas de tela (pixels) para NDC e reposiciona objeto."""
    global position_x, position_y

    if button != GLUT_LEFT_BUTTON or state != GLUT_DOWN:
        return

    width = glutGet(GLUT_WINDOW_WIDTH)
    height = glutGet(GLUT_WINDOW_HEIGHT)
    position_x = 2.0 * (x / width) - 1.0
    position_y = 1.0 - 2.0 * (y / height)
    glutPostRedisplay()


def main() -> None:
    """Entry point da aula 04."""
    glutInit()
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB)
    glutInitWindowSize(900, 600)
    glutCreateWindow(b"Aula 04 - Teclado e Mouse")
    glClearColor(0.07, 0.07, 0.1, 1.0)

    glutDisplayFunc(display)
    glutKeyboardFunc(keyboard)
    glutMouseFunc(mouse)

    print("Use WASD, RGB, clique esquerdo e ESC.")
    glutMainLoop()


if __name__ == "__main__":
    main()

File: glut_opengl_estudos/04_entrada_teclado_mouse.cpp
/*
Aula 04 — Interação com teclado e mouse (C++)

Conceitos:
- Callback de teclado (`glutKeyboardFunc`).
- Callback de mouse (`glutMouseFunc`).
- Estado global simples para aplicações interativas.

Controles:
- WASD: move o quadrado.
- R/G/B: altera cor.
- Clique esquerdo: reposiciona no ponto clicado.
- ESC: sair.
*/

#include <GL/glut.h>
#include <cstdlib>
#include "cpp_glut_utils.hpp"

namespace {

float positionX = 0.0f;
float positionY = 0.0f;
float colorR = 0.9f;
float colorG = 0.4f;
float colorB = 0.2f;

void drawSquare() {
    glBegin(GL_QUADS);
    glVertex2f(-0.15f, -0.15f);
    glVertex2f( 0.15f, -0.15f);
    glVertex2f( 0.15f,  0.15f);
    glVertex2f(-0.15f,  0.15f);
    glEnd();
}

void display() {
    glClear(GL_COLOR_BUFFER_BIT);
    glLoadIdentity();

    glTranslatef(positionX, positionY, 0.0f);
    glColor3f(colorR, colorG, colorB);
    drawSquare();

    glFlush();
}

void keyboard(unsigned char key, int, int) {
    constexpr float step = 0.05f;

    switch (key) {
        case 'w': case 'W': positionY += step; break;
        case 's': case 'S': positionY -= step; break;
        case 'a': case 'A': positionX -= step; break;
        case 'd': case 'D': positionX += step; break;
        case 'r': case 'R': colorR = 1.0f; colorG = 0.2f; colorB = 0.2f; break;
        case 'g': case 'G': colorR = 0.2f; colorG = 1.0f; colorB = 0.2f; break;
        case 'b': case 'B': colorR = 0.2f; colorG = 0.4f; colorB = 1.0f; break;
        case 27: std::exit(0); // ESC
    }

    glutPostRedisplay();
}

void mouse(int button, int state, int x, int y) {
    if (button != GLUT_LEFT_BUTTON || state != GLUT_DOWN) return;

    // Converte pixel para NDC em [-1, 1].
    const int width = glutGet(GLUT_WINDOW_WIDTH);
    const int height = glutGet(GLUT_WINDOW_HEIGHT);

    positionX = 2.0f * (static_cast<float>(x) / static_cast<float>(width)) - 1.0f;
    positionY = 1.0f - 2.0f * (static_cast<float>(y) / static_cast<float>(height));

    glutPostRedisplay();
}

} // namespace

int main(int argc, char** argv) {
    glutInit(&argc, argv);
    aula::configureCloseBehavior();
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB);
    glutInitWindowSize(900, 600);
    glutCreateWindow("Aula 04 - Teclado e Mouse (C++)");

    glClearColor(0.07f, 0.07f, 0.1f, 1.0f);
    glutDisplayFunc(display);
    glutKeyboardFunc(keyboard);
    glutMouseFunc(mouse);

    glutMainLoop();
    return 0;
}
